### ISSUE TITLE
CameraManager latitude must be between -90 and 90

### ISSUE THREAD

**Issue by kmadsen** [mapbox/mapbox-navigation-android#4632]
Capturing a crash that is coming from robo runs on this dev branch https://github.com/mapbox/mapbox-navigation-android/pull/4601

`com.mapbox.maps:android:10.0.0-rc.3`

https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.2e74b6b076d6fa53/matrices/6972915264876338818/executions/bs.52a762de69bc61a7
```
java.lang.Error: latitude must be between -90 and 90
     FATAL EXCEPTION: Thread-1149
Process: com.mapbox.navigation.examples, PID: 24356
java.lang.Error: latitude must be between -90 and 90
	at com.mapbox.maps.CameraManager.setCamera(Native Method)
	at com.mapbox.maps.NativeMapImpl.setCamera()
	at com.mapbox.maps.MapboxMap$setCamera$1.invoke()
	at com.mapbox.maps.MapboxMap$setCamera$1.invoke()
	at com.mapbox.maps.UtilsKt.call()
	at com.mapbox.maps.MapboxMap.setCamera()
	at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.performMapJump()
	at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.access$performMapJump()
	at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$registerInternalUpdateListener$1.onAnimationUpdate()
	at android.animation.ValueAnimator.animateValue(ValueAnimator.java:1463)
	at android.animation.ValueAnimator.animationFrame(ValueAnimator.java:1384)
	at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1427)
	at android.animation.ValueAnimator$AnimationHandler.doAnimationFrame(ValueAnimator.java:759)
	at android.animation.ValueAnimator$AnimationHandler$1.run(ValueAnimator.java:801)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
	at android.view.Choreographer.doFrame(Choreographer.java:603)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:10)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:7)
	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:17)
	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:3)
	at androidx.test.espresso.ViewInteraction.access$100(ViewInteraction.java:1)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:2)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by kiryldz** [comment]
No ideas so far sadly.
Is there any chance to reproduce?

**Comment by kmadsen** [comment]
> Is there any chance to reproduce?

So far I'm not reproducing locally. Will keep trying

Here's another one that is similar. 

https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.2e74b6b076d6fa53/matrices/6099964951520577164/executions/bs.b492c5646ebed85e
```
java.lang.Error: latitude must be between -90 and 90
     FATAL EXCEPTION: MapboxRenderThread
Process: com.mapbox.navigation.examples, PID: 27926
java.lang.Error: latitude must be between -90 and 90
	at com.mapbox.maps.Map.render(Native Method)
	at com.mapbox.maps.NativeMapImpl.render()
	at com.mapbox.maps.renderer.MapboxRenderer.onDrawFrame()
	at com.mapbox.maps.renderer.MapboxRenderThread.draw()
	at com.mapbox.maps.renderer.MapboxRenderThread.doFrame()
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:856)
	at android.view.Choreographer.doCallbacks(Choreographer.java:670)
	at android.view.Choreographer.doFrame(Choreographer.java:603)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by kiryldz** [comment]
@kmadsen could you please also specify maps sdk version that is used?

**Comment by kmadsen** [comment]
> @ kmadsen could you please also specify maps sdk version that is used?

Yep! `10.0.0-rc.3`. I'll update the description as I get more info as well

**Comment by kmadsen** [comment]
@kiryldz I can imagine this is a latitude/longitude swap somewhere. 

Can we update the log, to include the value that is causing the crash "latitude must be between -90 and 90, but was xxx.xxx"

**Comment by kiryldz** [comment]
> @kiryldz I can imagine this is a latitude/longitude swap somewhere.
> 
> Can we update the log, to include the value that is causing the crash "latitude must be between -90 and 90, but was xxx.xxx"

We could but it should happen in bindgen afair, so it would take some time (quite a long time 😆) to integrate it here. I'll check it out tomorrow and try to start that process because some of the errors coming from core are indeed not very informative 😄

**Comment by kmadsen** [comment]
Found another way to reproduce this. Am testing with an emulator and you can get a better stack trace

```
2021-07-21 17:28:35.236 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err: java.lang.Error: latitude must be between -90 and 90
    
    --------- beginning of crash
2021-07-21 17:28:35.246 5739-5803/com.mapbox.navigation.instrumentation_tests E/AndroidRuntime: FATAL EXCEPTION: MapboxRenderThread
    Process: com.mapbox.navigation.instrumentation_tests, PID: 5739
    java.lang.Error: latitude must be between -90 and 90
        at com.mapbox.maps.Map.render(Native Method)
        at com.mapbox.maps.NativeMapImpl.render(NativeMapImpl.kt:31)
        at com.mapbox.maps.renderer.MapboxRenderer.onDrawFrame(MapboxRenderer.kt:79)
        at com.mapbox.maps.renderer.MapboxRenderThread.draw(MapboxRenderThread.kt:169)
        at com.mapbox.maps.renderer.MapboxRenderThread.doFrame(MapboxRenderThread.kt:296)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:970)
        at android.view.Choreographer.doCallbacks(Choreographer.java:796)
        at android.view.Choreographer.doFrame(Choreographer.java:727)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:957)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.os.HandlerThread.run(HandlerThread.java:67)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.CameraManager.cameraForCoordinates(Native Method)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.NativeMapImpl.cameraForCoordinates(NativeMapImpl.kt:99)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.MapboxMap$cameraForCoordinates$1.invoke(MapboxMap.kt:454)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.MapboxMap$cameraForCoordinates$1.invoke(MapboxMap.kt:38)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.UtilsKt.call(Utils.kt:9)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.maps.MapboxMap.cameraForCoordinates(MapboxMap.kt:454)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.ui.maps.camera.data.MapboxNavigationViewportDataSource.updateOverviewData(MapboxNavigationViewportDataSource.kt:794)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.ui.maps.camera.data.MapboxNavigationViewportDataSource.evaluate(MapboxNavigationViewportDataSource.kt:335)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.instrumentation_tests.ui.SimpleMapViewNavigationTest$addNavigationCamera$1$3.onNewMapMatcherResult(SimpleMapViewNavigationTest.kt:144)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.core.trip.session.MapboxTripSession.updateMapMatcherResult(MapboxTripSession.kt:598)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.core.trip.session.MapboxTripSession.access$updateMapMatcherResult(MapboxTripSession.kt:60)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.mapbox.navigation.core.trip.session.MapboxTripSession$navigatorObserver$1.onStatus(MapboxTripSession.kt:212)
2021-07-21 17:28:35.247 5739-5864/com.mapbox.navigation.instrumentation_tests W/valhalla: No metadata found for tiles
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.os.MessageQueue.nativePollOnce(Native Method)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.os.MessageQueue.next(MessageQueue.java:335)
2021-07-21 17:28:35.247 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at androidx.test.espresso.base.Interrogator.getNextMessage(Interrogator.java:199)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:140)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.os.Handler.handleCallback(Handler.java:938)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:99)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.os.Looper.loop(Looper.java:223)
2021-07-21 17:28:35.250 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7656)
2021-07-21 17:28:35.251 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2021-07-21 17:28:35.251 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
2021-07-21 17:28:35.255 5739-5739/com.mapbox.navigation.instrumentation_tests W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
2021-07-21 17:28:35.266 628-5875/system_process I/DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
2021-07-21 17:28:35.268 5739-5739/com.mapbox.navigation.instrumentation_tests A/libc: /buildbot/src/android/ndk-release-r21/external/libcxx/../../external/libcxxabi/src/abort_message.cpp:72: abort_message: assertion "terminating with uncaught exception of type jni::PendingJavaException" failed
2021-07-21 17:28:35.268 5739-5739/com.mapbox.navigation.instrumentation_tests A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 5739 (mentation_tests), pid 5739 (mentation_tests)
2021-07-21 17:28:35.284 5739-5864/com.mapbox.navigation.instrumentation_tests W/valhalla: No metadata found for tiles
2021-07-21 17:28:35.303 5739-5803/com.mapbox.navigation.instrumentation_tests I/Process: Sending signal. PID: 5739 SIG: 9
2021-07-21 17:28:35.309 628-658/system_process W/BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.stats.service.DropBoxEntryAddedReceiver
2021-07-21 17:28:35.309 628-658/system_process W/BroadcastQueue: Background execution not allowed: receiving Intent { act=android.intent.action.DROPBOX_ENTRY_ADDED flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$PersistentTrustedReceiver
```

**Comment by tobrun** [comment]
@kmadsen 

For that given stacktrace, the issue is in nav-native:

> 179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000

> 179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000 

These aren't valid latitude / longitude values.


---------------------------------------


@kiryldz the crash happens in gl-native c++, if we give them invalid data, they crash. There is no clear path for them to actually mutate the camera. What if we do some sanity validation on the input we are giving to gl-native. If invalid, we ignore that update?

**Comment by kiryldz** [comment]
@tobrun I don't agree here. Passing invalid lat / lng values to `Point` imho should cause a crash and it does not matter much where do we throw exception from (from our SDK or gl-native).
I've prepared https://github.com/mapbox/mapbox-gl-native-internal/pull/2042 so that we will not get more informative error text. 
The cause of the issue raised here is actually sending invalid latitude. However root cause (where does it come from) is unclear. Obviously if we prepare invalid `Point` explicitly and pass it to gl-native - we would get such crash but main problem here is actually why did we do that.

**Comment by pengdev** [comment]
@kiryldz @tobrun I think the Point actually comes from the geojson dependency, it's not a generated class, and we could probably add the sanity check in the geojson dependency.

**Comment by kiryldz** [comment]
> @kiryldz @tobrun I think the Point actually comes from the geojson dependency, it's not a generated class, and we could probably add the sanity check in the geojson dependency.

I know that it comes from there 😄 All that I'm saying that if incorrect lat / lng values are provided - imho it must be a crash (and it works that way now). Our main goal now is try to understand __why__ we actually get it in certain scenarios.

**Comment by tobrun** [comment]
> why? 

-> https://github.com/mapbox/mapbox-navigation-native/issues/3634 :homer:

**Comment by kiryldz** [comment]
> > why?
> 
> -> [mapbox/mapbox-navigation-native#3634](https://github.com/mapbox/mapbox-navigation-native/issues/3634) :homer:

Haha, I actually thought @kmadsen just put mega-large parameters explicitly to `Point` and caused a crash 🤦 
That's nice that we know that such numbers go from nav-native in some cases (that are reproducable) 👍

**Comment by tobrun** [comment]
re. @pengdev 

>  sanity check in the geojson dependency.

The GeoJSON spec doesn't enforce position value boundaries:
 - https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.1

For this reason, I don't think we should add such a sanity check in the library. 
From a GeoJSON spec point-of-view, the provided Point latitude/longitude is valid. 

If we are in agreement that the issue is with the data provided by navigation and there is no tailwork to be done from our end (works-as-designed). I would be in favor of closing this ticket.

**Comment by kmadsen** [comment]
> If we are in agreement that the issue is with the data provided by navigation and there is no tailwork to be done from our end (works-as-designed). I would be in favor of closing this ticket.

This ticket is still representing an issue with `mapbox-navigation-android` instrumentation tests, which is why I'd like to keep it open until the issue is resolved. The root cause may be fixed by mapbox/mapbox-navigation-native#36

Also agree and can see that maps is working as expected. Although, it would be helpful if the value of the offender was in the logs: e.g., "java.lang.Error: latitude must be between -90 and 90 but was 179769313486231570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000"

**Comment by kmadsen** [comment]
I'm not reproducing anymore. Closing this as not reproducible. Can open a new investigation if we see it. 

And if we are reproducing, helpful to capture a nav-native history file
